framework/sandstone.cpp

```diff
@@ -2160,8 +2160,15 @@ static TestResult run_one_test_once(const struct test *test)
         (void) missing;
 
         children.results.emplace_back(ChildExitStatus{ TestResult::Skipped });
-    } else {
+    } else if (test->quality_level >= 0 && test->quality_level < sApp->requested_quality) {
+        init_per_thread_data();
+        log_skip(TestResourceIssueSkipCategory, "Test %s is in BETA quality, try again using --beta option", test->id);
+        children.results.emplace_back(ChildExitStatus{ TestResult::Skipped });
+    } else if (test->quality_level >= sApp->requested_quality) {
         run_one_test_children(children, test);
+    } else {
+        init_per_thread_data();
+        children.results.emplace_back(ChildExitStatus{ TestResult::Skipped });
```

Comment:
Do we still reach this at any time? The `get_first_test` and `next_text` functions below will skip the test if the quality is negative and the requested quality non-negative.